Strange extra type error after missing semicolon error

Given the following code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=55059b7460a49a908e086ce00bf462e2)):

```rust
fn f() -> f32 {
    3
    5.0
}

fn main() {
}
```

The current output is:

```
error: expected `;`, found `5.0`
 --> epp.rs:2:6
  |
2 |     3
  |      ^ help: add `;` here
3 |     5.0
  |     --- unexpected token

error[E0308]: mismatched types
 --> epp.rs:2:5
  |
2 |     3
  |     ^ expected `()`, found integer
```

What is this second error trying to tell me? _What_ did it expect to be unit type and why? Does it make sense for this error to be emitted despite the first error correctly identifying my problem?

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"chenyukang"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

That theory doesn't seem right to me, for two reasons:

* I would expect eating the `5.0` to result in a type error like "expected `f32` because of return type", but the actual error involves integer and unit.
* Repeating the missing-semi error results in repeating the type error as well, even though it's all within one function. [For example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=de50d5b7a5ca47a6fd6025cdae2a7000):

```rust
fn f() -> f32 {
    2_u32
    3_i8
    5.0
}

fn main() {
}
```

gives two missing-semi errors and two type errors:

```
error: expected `;`, found `3_i8`
 --> src/lib.rs:2:10
  |
2 |     2_u32
  |          ^ help: add `;` here
3 |     3_i8
  |     ---- unexpected token

error: expected `;`, found `5.0`
 --> src/lib.rs:3:9
  |
3 |     3_i8
  |         ^ help: add `;` here
4 |     5.0
  |     --- unexpected token

error[E0308]: mismatched types
 --> src/lib.rs:2:5
  |
2 |     2_u32
  |     ^^^^^ expected `()`, found `u32`

error[E0308]: mismatched types
 --> src/lib.rs:3:5
  |
3 |     3_i8
  |     ^^^^ expected `()`, found `i8`
```

--- a/compiler/rustc_parse/src/parser/stmt.rs
+++ b/compiler/rustc_parse/src/parser/stmt.rs
@@ -553,39 +553,46 @@ impl<'a> Parser<'a> {
         match stmt.kind {
             // Expression without semicolon.
             StmtKind::Expr(ref mut expr)
-                if self.token != token::Eof && classify::expr_requires_semi_to_be_stmt(expr) =>
-            {
+                if self.token != token::Eof && classify::expr_requires_semi_to_be_stmt(expr) => {
                 // Just check for errors and recover; do not eat semicolon yet.
-                if let Err(mut e) =
-                    self.expect_one_of(&[], &[token::Semi, token::CloseDelim(Delimiter::Brace)])
-                {
-                    if let TokenKind::DocComment(..) = self.token.kind {
-                        if let Ok(snippet) = self.span_to_snippet(self.token.span) {
-                            let sp = self.token.span;
-                            let marker = &snippet[..3];
-                            let (comment_marker, doc_comment_marker) = marker.split_at(2);
+                // `expect_one_of` returns PResult<'a, bool /* recovered */>
+                let replace_with_err =
+                    match self.expect_one_of(&[], &[token::Semi, token::CloseDelim(Delimiter::Brace)]) {
+                    // Recover from parser, skip type error to avoid extra errors.
+                    Ok(true) => true,
+                    Err(mut e) => {
+                        if let TokenKind::DocComment(..) = self.token.kind &&
+                            let Ok(snippet) = self.span_to_snippet(self.token.span) {
+                                let sp = self.token.span;
+                                let marker = &snippet[..3];
+                                let (comment_marker, doc_comment_marker) = marker.split_at(2);
 
-                            e.span_suggestion(
-                                sp.with_hi(sp.lo() + BytePos(marker.len() as u32)),
-                                &format!(
-                                    "add a space before `{}` to use a regular comment",
-                                    doc_comment_marker,
-                                ),
-                                format!("{} {}", comment_marker, doc_comment_marker),
-                                Applicability::MaybeIncorrect,
-                            );
+                                e.span_suggestion(
+                                    sp.with_hi(sp.lo() + BytePos(marker.len() as u32)),
+                                    &format!(
+                                        "add a space before `{}` to use a regular comment",
+                                        doc_comment_marker,
+                                    ),
+                                    format!("{} {}", comment_marker, doc_comment_marker),
+                                    Applicability::MaybeIncorrect,
+                                );
                         }
-                    }
-                    if let Err(mut e) =
-                        self.check_mistyped_turbofish_with_multiple_type_params(e, expr)
-                    {
-                        if recover.no() {
-                            return Err(e);
+
+                        if let Err(mut e) =
+                            self.check_mistyped_turbofish_with_multiple_type_params(e, expr)
+                        {
+                            if recover.no() {
+                                return Err(e);
+                            }
+                            e.emit();
+                            self.recover_stmt();
                         }
-                        e.emit();
-                        self.recover_stmt();
+                        true
                     }
-                    // Don't complain about type errors in body tail after parse error (#57383).
+                    _ => false
+                };
+                if replace_with_err {
+                    // We already emitted an error, so don't emit another type error
                     let sp = expr.span.to(self.prev_token.span);
                     *expr = self.mk_expr_err(sp);
                 }

--- a/src/test/ui/parser/issue-103425.rs
+++ b/src/test/ui/parser/issue-103425.rs
@@ -1,0 +1,15 @@
+fn f() -> f32 {
+    3
+    //~^ ERROR expected `;`
+    5.0
+}
+
+fn k() -> f32 {
+    2_u32
+    //~^ ERROR expected `;`
+    3_i8
+    //~^ ERROR expected `;`
+    5.0
+}
+
+fn main() {}

--- a/src/test/ui/parser/issue-103425.stderr
+++ b/src/test/ui/parser/issue-103425.stderr
@@ -1,0 +1,29 @@
+error: expected `;`, found `5.0`
+  --> $DIR/issue-103425.rs:2:6
+   |
+LL |     3
+   |      ^ help: add `;` here
+LL |
+LL |     5.0
+   |     --- unexpected token
+
+error: expected `;`, found `3_i8`
+  --> $DIR/issue-103425.rs:8:10
+   |
+LL |     2_u32
+   |          ^ help: add `;` here
+LL |
+LL |     3_i8
+   |     ---- unexpected token
+
+error: expected `;`, found `5.0`
+  --> $DIR/issue-103425.rs:10:9
+   |
+LL |     3_i8
+   |         ^ help: add `;` here
+LL |
+LL |     5.0
+   |     --- unexpected token
+
+error: aborting due to 3 previous errors
+

